glean/src/server_events.rs

```diff
@@ -0,0 +1,309 @@
+    pub event: Option<Box<dyn EventsPingEvent>>, // valid event of `EventsPingEvent` for this ping.
+}
+
+// Record and submit `events` ping
+impl GleanEventsLogger {
+    /// General `record_events_ping` function
+    pub fn record_events_ping(&self, request_info: &RequestInfo, params: &EventsPing) {
+        // Define the outer `Metrics` map that holds the metric type.
+        let mut metrics = Metrics::new();
+        // Create the inner metric value map to insert into `Metrics`.
+        // Create corresponding metric value maps to insert into `Metrics`.
+        let mut string_map: HashMap<String, serde_json::Value> = std::collections::HashMap::new();
+        string_map.insert(
+            "syncstorage.device_family".to_string(),
+            serde_json::Value::String(params.syncstorage_device_family.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.hashed_device_id".to_string(),
+            serde_json::Value::String(params.syncstorage_hashed_device_id.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.hashed_fxa_uid".to_string(),
+            serde_json::Value::String(params.syncstorage_hashed_fxa_uid.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.platform".to_string(),
+            serde_json::Value::String(params.syncstorage_platform.to_string()),
+        );
+        metrics.insert("string".to_string(), string_map);
+
+        let mut events: Vec<GleanEvent> = Vec::new();
+        if let Some(event) = &params.event {
+            events.push(event.glean_event());
+        }
+
+        self.record("events", request_info, metrics, events);
+    }
+}
+
+// Record and submit `events` ping omitting user request info
+impl GleanEventsLogger {
+    pub fn record_events_ping_without_user_info(&self, params: &EventsPing) {
+        self.record_events_ping(&RequestInfo::default(), params)
+    }
+}
+
+pub struct SyncDauPing {
+    pub syncstorage_device_family: String, // Device family from which sync action was initiated. Desktop PC, Tablet, Mobile, and Other.
+    pub syncstorage_hashed_device_id: String, // Hashed device id that is associated with a given account.
+    pub syncstorage_hashed_fxa_uid: String, // User identifier. Uses `hashed_fxa_uid` for accurate count of sync actions.
+    pub syncstorage_platform: String, // Platform from which sync action was initiated. Firefox Desktop, Fenix, or Firefox iOS.
+}
+
+// Record and submit `sync-dau` ping
+impl GleanEventsLogger {
+    /// General `record_events_ping` function
+    pub fn record_sync_dau_ping(&self, request_info: &RequestInfo, params: &SyncDauPing) {
+        // Define the outer `Metrics` map that holds the metric type.
+        let mut metrics = Metrics::new();
+        // Create the inner metric value map to insert into `Metrics`.
+        metrics.insert(
+            "string".to_string(),
```

Comment:
a bit nitty, but we generally call `to_owned()` instead of `to_string()`. Both are equally performant. `to_owned()` reflects that we're converting a static string reference (that's what "string" is) to a own-able String object. Since `insert(...)` essentially is consuming these values, it's semantically more correct.